src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Services/SemanticFacts/CSharpSemanticFacts.cs

```diff
@@ -57,23 +57,52 @@ public bool CanReplaceWithRValue(SemanticModel semanticModel, [NotNullWhen(true)
+            var tokenParent = token.Parent;
+            var sourceAncestor = tokenParent;
+            // Skip the tuple expression if the identifier implies the name of the tuple field
+            // For example, we are evaluating `x` in `(x, y)`, but not in `(x: x, y: y)`
+            if (tokenParent is IdentifierNameSyntax { Parent: ArgumentSyntax { Parent: TupleExpressionSyntax parentTuple } })
+            {
+                sourceAncestor = parentTuple.Parent;
+            }
+
+            if (sourceAncestor is null)
+                return null;
+
+            foreach (var ancestor in sourceAncestor.AncestorsAndSelf())
             {
                 var symbol = semanticModel.GetDeclaredSymbol(ancestor, cancellationToken);
                 if (symbol != null)
                 {
-                    if (symbol is IMethodSymbol { MethodKind: MethodKind.Conversion })
-                    {
-                        // The token may be part of a larger name (for example, `int` in `public static operator int[](Goo g);`.
-                        // So check if the symbol's location encompasses the span of the token we're asking about.
-                        if (symbol.Locations.Any(static (loc, location) => loc.SourceTree == location.SourceTree && loc.SourceSpan.Contains(location.SourceSpan), location))
-                            return symbol;
-                    }
-                    else
+                    switch (symbol)
                     {
-                        // For any other symbols, we only care if the name directly matches the span of the token
-                        if (symbol.Locations.Contains(location))
-                            return symbol;
+                        case IMethodSymbol { MethodKind: MethodKind.Conversion }:
+                            {
+                                // The token may be part of a larger name (for example, `int` in `public static operator int[](Goo g);`.
+                                // So check if the symbol's location encompasses the span of the token we're asking about.
+                                if (symbol.Locations.Any(static (loc, location) => loc.SourceTree == location.SourceTree && loc.SourceSpan.Contains(location.SourceSpan), location))
+                                    return symbol;
+                                break;
+                            }
+
+                        case IPropertySymbol { ContainingType: ITypeSymbol { IsAnonymousType: true } }:
+                            {
+                                // If the span is not part of the property's explicit name, it's not our target result
+                                // For example, we are evaluating `a.Length` in `new { a.Length }`
+                                if (ancestor is AnonymousObjectMemberDeclaratorSyntax declarator &&
+                                    (declarator.NameEquals is null || !declarator.NameEquals.Span.Contains(location.SourceSpan)))
```

Comment:
i don't understand this check tbh.  if the sourceSpan was not within the declarator, then we would have never hit hte NameEquals in the first place, right?

Reply:
Correct, your original source span must be within the declarator to also match the span of the NameEquals syntax. I don't understand what you're concerned about though, perhaps the ancestor traversal would've already terminated before reaching a declarator syntax?

Reply:
aren't both of the checks actually about being in the span of the initializer of the declarator?  It looks like you're saying "it can't be in the name portion".  But instead of checking the negative, check the positive.